odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1160 @@
+    is the real numbers. It is possible to create functions that do not fulfil
+    these assumptions, however some mathematical results might not be valide in
+    this case. For more information, see
+    http://odl.readthedocs.io/guide/in_depth/functional_guide.html.
+    """
+
+    def __init__(self, space, linear=False, grad_lipschitz=np.inf):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        space : `LinearSpace`
+            The domain of this functional, i.e., the set of elements to
+            which this functional can be applied.
+        linear : bool, optional
+            If `True`, the functional is considered as linear.
+        grad_lipschitz : float, optional
+            The Lipschitz constant of the gradient.
+        """
+        Operator.__init__(self, domain=space,
+                          range=space.field, linear=linear)
+        self.__grad_lipschitz = float(grad_lipschitz)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        raise NotImplementedError
+
+    @property
+    def proximal(self):
+        """Return the proximal factory of the functional.
+
+        Notes
+        -----
+        The nonsmooth solvers that make use of proximal operators to solve a
+        given optimization problem take a `proximal factory` as input,
+        i.e., a function returning a proximal operator. Note that
+        ``Functional.proximal`` is in fact a `proximal factory`. See for
+        example `forward_backward_pd`.
+        """
+        raise NotImplementedError
+
+    @property
+    def convex_conj(self):
+        """Convex conjugate functional of the functional.
+
+        Notes
+        -----
+        The convex conjugate functional of a convex functional :math:`f(x)`,
+        defined on a Hilber space, is defined as the functional
+
+        .. math::
+
+            f^*(x^*) = \\sup_{x} \{ \\langle x^*,x \\rangle - f(x)  \}.
+
+        The concept is also known as the Legendre transformation.
+
+        References
+        ----------
+        Wikipedia article on `Convex conjugate
+        <https://en.wikipedia.org/wiki/Convex_conjugate>`_.
+
+        Wikipedia article on `Legendre transformation
+        <https://en.wikipedia.org/wiki/Legendre_transformation>`_.
+
+        For literature references see, e.g., [Lue1969]_, [Roc1970]_.
+        """
+        raise NotImplementedError
+
+    def derivative(self, point):
+        """Return the derivative operator in the given point.
+
+        This function returns the linear operator given by
+
+            ``self.derivative(point)(x) == self.gradient(point).inner(x)``
```

Comment:
For places like this you can achieve the same effect (i think the typesetting actually improves somewhat) by

```
This function returns the linear operator given by::

    ``self.derivative(point)(x) == self.gradient(point).inner(x)``
```